### PR TITLE
Agda's 'generalized' feature

### LINKED ISSUE #27: Agda or Lean style generalization
As title. Agda has some technical slides, I think we could learn from that.

### PR DESCRIPTION
**Pull request by ice1000**
Fix #27, /cc @AustinZhu 

Summary:
+ `variable ID+ ':' expr`
+ Similar to Agda's semantics, but yet doesn't support introducing new generalizations in ctors and fields (but this should be allowed, unlike for levels which cannot parameterize ctors/fields)
+ Modified the redblack tree showcase with generalization
+ Doesn't yet support referencing a `variable` from another `variable`

bors d=imkiva

### CODE REVIEW

File: base/src/main/java/org/aya/concrete/resolve/visitor/StmtResolver.java [hunk 1]
```diff
@@ -32,9 +32,9 @@ static void resolveStmt(@NotNull Stmt stmt, @NotNull ResolveInfo info) {
     switch (stmt) {
       case Command.Module mod -> resolveStmt(mod.contents(), info);
       case Decl.DataDecl decl -> {
-        var signatureResolver = new ExprResolver(new ExprResolver.Options(true));
+        var signatureResolver = new ExprResolver(ExprResolver.LAX);
         var local = signatureResolver.resolveParams(decl.telescope, decl.ctx);
-        decl.telescope = local._1;
+        decl.telescope = local._1.prependedAll(signatureResolver.allowedGeneralizes().valuesView());
```
Comment (imkiva) [inline base/src/main/java/org/aya/concrete/resolve/visitor/StmtResolver.java:37]: I guess we can move `.prependedAll(signatureResolver.allowedGeneralizes().valuesView())` to `signatureResolver.resolveParams`

**Review by imkiva — APPROVED** [review]
bors r+

File: base/src/main/java/org/aya/concrete/resolve/visitor/StmtResolver.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/concrete/resolve/visitor/StmtResolver.java:37]: Related #207

Comment (ice1000) [reply base/src/main/java/org/aya/concrete/resolve/visitor/StmtResolver.java:37]: > I guess we can move `.prependedAll(signatureResolver.allowedGeneralizes().valuesView())` to `signatureResolver.resolveParams` 

Are you gonna do it?

Comment (imkiva) [reply base/src/main/java/org/aya/concrete/resolve/visitor/StmtResolver.java:37]: Maybe tomorrow

Comment (imkiva) [reply base/src/main/java/org/aya/concrete/resolve/visitor/StmtResolver.java:37]: You're so fast. I just pulled the code and realized that you have done it. 🤣

**Pull request merged**